Spigot-Server-Patches/0349-Use-ConcurrentHashMap-in-JsonList.patch

```diff
@@ -0,0 +1,118 @@
+@@ -1,6 +1,6 @@
+ package net.minecraft.server;
+ 
+-import com.google.common.collect.Lists;
++// import com.google.common.collect.Lists; // Paper
+ import com.google.common.collect.Maps;
+ import com.google.common.io.Files;
+ import com.google.gson.Gson;
+@@ -21,11 +21,13 @@ import java.io.Reader;
+ import java.lang.reflect.ParameterizedType;
+ import java.lang.reflect.Type;
+ import java.nio.charset.StandardCharsets;
+-import java.util.ArrayList;
++// import java.util.ArrayList; // Paper
+ import java.util.Collection;
+ import java.util.Iterator;
+ import java.util.List;
+-import java.util.Map;
++// import java.util.Map; // Paper
++import java.util.concurrent.ConcurrentMap;
++
+ import org.apache.commons.io.IOUtils;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
+@@ -35,7 +37,10 @@ public class JsonList<K, V extends JsonListEntry<K>> {
+     protected static final Logger a = LogManager.getLogger();
+     protected final Gson b;
+     private final File c;
+-    private final Map<String, V> d = Maps.newHashMap();
++    // Paper start
++    // private final Map<String, V> d = Maps.newHashMap();
++    private final ConcurrentMap<String, V> d = Maps.newConcurrentMap();
++    // Paper end
+     private boolean e = true;
+     private static final ParameterizedType f = new ParameterizedType() {
+         public Type[] getActualTypeArguments() {
+@@ -83,8 +88,13 @@ public class JsonList<K, V extends JsonListEntry<K>> {
+     }
+ 
+     public V get(K k0) {
+-        this.h();
+-        return (V) this.d.get(this.a(k0)); // CraftBukkit - fix decompile error
++        // Paper start
++        // this.h();
++        // return (V) this.d.get(this.a(k0)); // CraftBukkit - fix decompile error
++        return (V) this.d.computeIfPresent(this.a(k0), (k, v) -> {
++            return v.hasExpired() ? null : v;
++        });
++        // Paper end
+     }
+ 
+     public void remove(K k0) {
+@@ -121,7 +131,8 @@ public class JsonList<K, V extends JsonListEntry<K>> {
+     }
+ 
+     private void h() {
+-        ArrayList arraylist = Lists.newArrayList();
++        // Paper start
++        /*ArrayList arraylist = Lists.newArrayList();
+         Iterator iterator = this.d.values().iterator();
+ 
+         while (iterator.hasNext()) {
+@@ -138,15 +149,22 @@ public class JsonList<K, V extends JsonListEntry<K>> {
+             Object object = iterator.next();
+ 
+             this.d.remove(object);
++        }*/
++        
++        for (Iterator<V> i = this.d.values().iterator(); i.hasNext();) {
++            if (i.next().hasExpired()) {
++                i.remove();
++            }
+         }
+-
++        // Paper end
+     }
+ 
+     protected JsonListEntry<K> a(JsonObject jsonobject) {
+         return new JsonListEntry((Object) null, jsonobject);
+     }
+ 
+-    protected Map<String, V> e() {
++    //protected Map<String, V> e() { // Paper
++    protected ConcurrentMap<String, V> e() { // Paper
```

Comment:
I don't really see the point of changing the return type here (and the same applies to the field): a `ConcurrentMap` is a `Map`, so unless you use a specific method that only exists in `ConcurrentMap`, you can just keep all this as-is and only change the initializer.

Reply:
yeah please remove this change

Reply:
Ah, yeah, the point of that change is to allow a later patch or plugin to use methods specific to ConcurrentMap if needed. I figured the change would actually be safe as anything that requested a from the function would be automatically typecast by the JVM.